### ISSUE TITLE
Question about loss D

### ISSUE THREAD

**Issue by sarrbranka** [odegeasslbc/FastGAN-pytorch#28]
Why is the loss D for fake images min(0,-1-d(x_fake)) not min(0,-d(x_fake))?
I feel like min(0,-1-d(x_fake)) can try to converge to -1 when the D is fed fake images.
I thought the ideal output of D from real images to be 1 and from fake images to be 0.
![Screenshot from 2021-08-09 10-29-25](https://user-images.githubusercontent.com/76247476/128652511-d92bed44-1c72-4c61-9fd5-494a8fbfacc3.png)

**Comment by odegeasslbc** [comment]
Hi. I think the statement "the ideal output of D from real images to be 1 and from fake images to be 0" is not a necessity. For a discriminator, all we hope it can do is to distinguish the real/fake images. I.e. D's output value does not need to fall into a certain range, as long as D gives real images with a score towards one direction while gives fake images a score towards an opposite direction, then it's fine. The two values can be [-1,1], can be [0,1], also can be [-inf, inf]. It simply depends on what equation you use to wrap the idea of the loss. For example, if you wrap a sigmoid function, it of course restricts the D output to [0, 1].

**Comment by LearningJack** [comment]
Could you help me understand the relationship with the code **err = F.relu( torch.rand_like(pred) * 0.2 + 0.8 + pred).mean()** and the d loss in the paper?
![image](https://user-images.githubusercontent.com/38984606/131849172-2e1c882d-9db4-4f00-9353-e4334bf71432.png)

**Comment by odegeasslbc** [comment]
1. Its a soft version of hinge loss. By applying randomness, the real samples are not treated as 1, but a value between 0.8 to 1. Please refer to ideas related to "label smoothing".
2. The core idea for hinge loss is: D is no longer be optimized when if performs good enough.
